contrib/epee/include/net/net_helper.h

```diff
@@ -165,23 +174,37 @@ namespace net_utils
 
 				boost::system::error_code ec = boost::asio::error::would_block;
 
-				m_ssl_socket.next_layer().async_connect(remote_endpoint, boost::lambda::var(ec) = boost::lambda::_1);
+				m_ssl_socket->next_layer().async_connect(remote_endpoint, boost::lambda::var(ec) = boost::lambda::_1);
 				while (ec == boost::asio::error::would_block)
 				{	
 					m_io_service.run_one(); 
 				}
 				
-				if (!ec && m_ssl_socket.next_layer().is_open())
+				if (!ec && m_ssl_socket->next_layer().is_open())
 				{
 					m_connected = true;
 					m_deadline.expires_at(std::chrono::steady_clock::time_point::max());
 					// SSL Options
-					if(m_ssl) {
-						// Disable verification of host certificate
-						m_ssl_socket.set_verify_mode(boost::asio::ssl::verify_peer);
-						// Handshake
-						m_ssl_socket.next_layer().set_option(boost::asio::ip::tcp::no_delay(true));
-						m_ssl_socket.handshake(boost::asio::ssl::stream_base::client);
+					if (m_ssl_support == epee::net_utils::e_ssl_support_enabled || m_ssl_support == epee::net_utils::e_ssl_support_autodetect)
+					{
+						if (!ssl_handshake(*m_ssl_socket, boost::asio::ssl::stream_base::client, m_ctx))
+						{
+							if (m_ssl_support == epee::net_utils::e_ssl_support_autodetect)
+							{
+								MERROR("SSL handshake failed on an autodetect connection, reconnecting without SSL");
+								m_ssl_support = epee::net_utils::e_ssl_support_disabled;
+								boost::system::error_code ignored_ec;
+								m_ssl_socket->next_layer().shutdown(boost::asio::ip::tcp::socket::shutdown_both, ignored_ec);
+								m_ssl_socket->next_layer().close();
+								goto again;
```

Comment:
Well, doing it this way seems fairly unobtrusive. A client may not know if the daemon (typically) it's connecting to supports SSL (for instance for public RPC dameons). This check means it can use SSL when the daemon supports it (granted, it doesn't protect at all against active attackers, but it does against passive ones).